### PR TITLE
Adds netlink and mqtt implementation for worker

### PR DESCRIPTION
**Pull request by awlx**
This adds a simple mqtt implementation for the worker.

The worker subscribes to `/wireguard/$domain/+`. And thus gets the information for which domain a key is received. 

Then it starts it's process to add corresponding routes, vxlan bridge fdb entries and wireguard keys via netlink.

### CODE REVIEW

**Comment by mweinelt** [comment]
Can you please describe the MQTT protocol that you dreamed up?

**Comment by awlx** [comment]
At the moment it's just get the key via `/wireguard/domain/+` then parse the domain part and go from there.

File: wgkex/worker/mqtt.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    client.connect(broker_address)
+    for domain in domains:
+        print("Subscribing to topic", "wireguard/" + domain + "/+")
+        client.subscribe("wireguard/" + domain + "/+")
+    client.loop_forever()
+
+
+def on_message(client, userdata, message):
+
+    domain = re.search("/.*ffmuc_(\w+)/", message.topic).group(1)
```
Comment (mweinelt) [inline wgkex/worker/mqtt.py:29]: Config value

File: wgkex/worker/mqtt.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+def connect(domains: str):
+    broker_address = "broker.ov.ffmuc.net"
+    client = mqtt.Client(socket.gethostname())
+    client.on_message = on_message
+    print("connecting to broker " + broker_address)
+    client.connect(broker_address)
+    for domain in domains:
+        print("Subscribing to topic", "wireguard/" + domain + "/+")
+        client.subscribe("wireguard/" + domain + "/+")
```
Comment (mweinelt) [inline wgkex/worker/mqtt.py:23]: If all workers subscribe tot the same generic path, how is the gateway ultimately selected?

File: wgkex/worker/mqtt.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    find_stale_wireguard_clients,
+    link_handler,
+    WireGuardClient,
+    generate_lladdr,
+    generate_interface_names,
+)
+
+
+def connect(domains: str):
+    broker_address = "broker.ov.ffmuc.net"
```
Comment (mweinelt) [inline wgkex/worker/mqtt.py:16]: Config value

File: wgkex/worker/netlink.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+# we receive stuff from wgkex-broker
+def generate_lladdr(public_key: str) -> str:
+    m = hashlib.md5()
+
+    m.update(public_key.encode("ascii") + b"\n")
+    hashed_key = m.hexdigest()
+    hash_as_list = wrap(hashed_key, 2)
+    temp_mac = ":".join(["02"] + hash_as_list[:5])
+
+    lladdr = re.sub("\/\d+$", "/128", mac2eui64(mac=temp_mac, prefix="fe80::/10"))
```
Comment (mweinelt) [inline wgkex/worker/netlink.py:45]: What happens here?

File: wgkex/worker/netlink.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+
+# pyroute2 stuff
+def link_handler(client: WireGuardClient) -> Dict:
+    results = {}
+
+    results.update({"Wireguard": wireguard_handler(client)})
+    try:
+        results.update({"Route": route_handler(client)})
+    except Exception as e:
```
Comment (mweinelt) [inline wgkex/worker/netlink.py:80]: Can we narrow the list of possible exception down to something less generic?

**Review by mweinelt — CHANGES_REQUESTED** [review]
I did a quick first pass, PTAL.

File: wgkex/worker/mqtt.py (hunk 2 above)
Comment (awlx) [reply wgkex/worker/mqtt.py:23]: The node selects the gateway at the moment. There is no feedback channel from the Broker.

File: wgkex/worker/netlink.py (hunk 4 above)
Comment (awlx) [reply wgkex/worker/netlink.py:45]: It replaces the /10 to /128.

File: wgkex/worker/netlink.py (hunk 5 above)
Comment (awlx) [reply wgkex/worker/netlink.py:80]: Good question, I mean we get file exists if we add the route when it's already there. Which is fine we just can keep on going.

File: wgkex/worker/netlink.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+    action = "append"
+    if client.remove:
+        action = "del"
+
+    return ip.fdb(
+        action,
+        ifindex=ip.link_lookup(ifname=client.vx_interface)[0],
+        lladdr="00:00:00:00:00:00",
+        dst=re.sub("\/\d+$", "", client.lladdr),
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply wgkex/worker/netlink.py:98]: I might have misunderstood what this re.sub call does, can you explain it?

File: wgkex/worker/netlink.py (hunk 4 above)
Comment (mweinelt) [reply wgkex/worker/netlink.py:45]: What happens when you pass a /128 as prefix to mac2eui64?

File: wgkex/worker/netlink.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+    action = "add"
+    if client.remove:
+        action = "del"
+
+    result = ip.route(
+        action, dst=client.lladdr, oif=ip.link_lookup(ifname=client.wg_interface)[0]
+    )
+
+    ip.close()
```
Comment (mweinelt) [inline wgkex/worker/netlink.py:129]:
```suggestion
    with IPRoute() as ip: 
        result = ip.route(
            "del" if client.remove else "add, dst=client.lladdr, oif=ip.link_lookup(ifname=client.wg_interface)[0]
        )
```

Comment (mweinelt) [reply wgkex/worker/netlink.py:129]: Saw you forgot to close the socket, so use context handlers instead.

File: wgkex/worker/netlink.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+    result = ip.fdb(
+        action,
+        ifindex=ip.link_lookup(ifname=client.vx_interface)[0],
+        lladdr="00:00:00:00:00:00",
+        dst=re.sub("\/\d+$", "", client.lladdr),
+    )
+
+    ip.close()
+    return(result)
```
Comment (moepman) [inline wgkex/worker/netlink.py:102]:
```suggestion
    return result
```
remove redundant parenthesis

File: wgkex/worker/netlink.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    return peer
+
+
+def cleanup_wireguard_clients(domain: str) -> bool:
+    stale_clients = find_stale_wireguard_clients("wg-" + domain)
+    result = []
+    for stale_client in stale_clients:
+        stale_wireguard_client = WireGuardClient(
+            public_key=stale_client,
+            lladdr=generate_lladdr(stale_client),
```
Comment (moepman) [inline wgkex/worker/netlink.py:62]: **Expected type 'bytes', got 'str' instead** -> either encode here or change generate_lladdr

File: wgkex/worker/netlink.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            domain=domain,
+            wg_interface="",
+            vx_interface="",
+            remove=True,
+        )
+        stale_wireguard_client = generate_interface_names(stale_wireguard_client)
+        result = link_handler(stale_wireguard_client)
+    return result
```
Comment (moepman) [inline wgkex/worker/netlink.py:70]: **Expected type 'bool', got 'Union[list, dict]' instead**

File: wgkex/worker/netlink.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    if client.remove:
+        action = "del"
+
+    result = ip.route(
+        action, dst=client.lladdr, oif=ip.link_lookup(ifname=client.wg_interface)[0]
+    )
+
+    ip.close()
+
+    return(result)
```
Comment (moepman) [inline wgkex/worker/netlink.py:131]:
```suggestion
    return result
```

File: wgkex/worker/netlink.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+        "persistent_keepalive": 15,
+        "allowed_ips": [client.lladdr],
+        "remove": client.remove,
+    }
+
+    return wg.set(client.wg_interface, peer=wg_peer)
+
+
+def route_handler(client: WireGuardClient) -> Dict:
+    with IPRoute() as ip:
```
Comment (moepman) [inline wgkex/worker/netlink.py:119]:
```suggestion
    with IPRoute() as ip:
```
remove whitespace

File: wgkex/worker/netlink.py [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    for stale_client in stale_clients:
+        stale_wireguard_client = WireGuardClient(
+            public_key=stale_client,
+            lladdr=generate_lladdr(stale_client),
+            domain=domain,
+            wg_interface="",
+            vx_interface="",
+            remove=True,
+        )
+        stale_wireguard_client = generate_interface_names(stale_wireguard_client)
```
Comment (moepman) [inline wgkex/worker/netlink.py:68]: **Unresolved reference 'generate_interface_names'**
```suggestion
        stale_wireguard_client = generate_ifname(stale_wireguard_client)
```

File: wgkex/worker/netlink.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+        "remove": client.remove,
+    }
+
+    return wg.set(client.wg_interface, peer=wg_peer)
+
+
+def route_handler(client: WireGuardClient) -> Dict:
+    with IPRoute() as ip: 
+        result = ip.route(
+            "del" if client.remove else "add, dst=client.lladdr, oif=ip.link_lookup(ifname=client.wg_interface)[0]
```
Comment (moepman) [inline wgkex/worker/netlink.py:121]: **Missing closing quote ["]**
```suggestion
            "del" if client.remove else "add", dst=client.lladdr, oif=ip.link_lookup(ifname=client.wg_interface)[0]
```

**Pull request merged**